torch/distributed/distributed_c10d.py

```diff
@@ -1247,6 +1265,10 @@ def _new_process_group_helper(
             "created, please use a different group name"
         )
 
+    if device_id is not None and (device_id.index is None or device_id.type != 'cuda'):
+        raise ValueError("init_process_group device_id parameter must be a cuda device with an "
+                         "id, e.g. cuda:0, not just cuda or cpu")
+
```

Comment:
My intention is someday we do for sure.  Also it may not be device specific but could be (device, backend) specific like cuda/rccl or cuda/msccl and not just cuda/nccl.  It feels like we're making some assumptions cuda==nccl when maybe we need to abstract something?

I'm not sure how best to do it right now or what the more generic interfaces might look like so I punted here for when we are plumbing those devices in more, or if we need to do it through a custom backend extension capability system or plugin or something.